### PR TITLE
Automatic determination of reg param

### PR DESCRIPTION
**Pull request by ezradlesser**
Has not been tested with grooming -- FYI there may be some inconsistencies with `obs_setting` and `obs_label` vars in `run_analysis.py` which should be tested further

### CODE REVIEW

File: pyjetty/alice_analysis/analysis/user/substructure/run_analysis.py [hunk 1]
```diff
@@ -9,25 +9,25 @@
     The grooming part is optional, and should be labeled e.g. zcut01_B0 — from CommonUtils::grooming_label({'sd':[zcut, beta]})
     For example: h_subjet_z_JetPt_R0.4_0.1
     For example: h_subjet_z_JetPt_R0.4_0.1_zcut01_B0
-    
+
```
Comment (jdmulligan) [inline pyjetty/alice_analysis/analysis/user/substructure/run_analysis.py:12]: A lot of whitespace changes here...is this some autoformat thing? Would be good if we can avoid spamming PRs with these ;)

File: pyjetty/alice_analysis/analysis/user/substructure/run_analysis.py [hunk 2]
```diff
@@ -174,75 +180,79 @@ def create_output_subdir(self, output_dir, name):
...
     for jetR in self.jetR_list:
-    
+
       # Loop through subconfigurations to unfold
       for i, _ in enumerate(self.obs_subconfig_list):
 
         obs_setting = self.obs_settings[i]
         grooming_setting = self.grooming_settings[i]
         obs_label = self.utils.obs_label(obs_setting, grooming_setting)
-        reg_param_final = self.utils.get_reg_param(self.obs_settings, self.grooming_settings, self.obs_subconfig_list, self.obs_config_dict, obs_label, jetR)
```
Comment (jdmulligan) [inline pyjetty/alice_analysis/analysis/user/substructure/run_analysis.py:190]: I think we should still keep this functionality -- to be able to manually override the automatically determined regularization parameter. Can you keep this in, such that if we specify a manual value in the config it uses that?

File: pyjetty/alice_analysis/analysis/user/substructure/run_analysis.py [hunk 3]
```diff
@@ -174,75 +180,79 @@ def create_output_subdir(self, output_dir, name):
...
-      if systematic == 'truncation':
+      elif systematic == 'truncation':
         truncation = True
-      if systematic == 'binning':
+      elif systematic == 'binning':
         binning = True
-      
-      analysis = roounfold_obs.Roounfold_Obs(self.observable, data, response, self.config_file, output_dir, self.file_format, rebin_response=rebin_response, prior_variation_parameter=prior_variation_parameter, truncation=truncation, binning=binning)
+
+      analysis = roounfold_obs.Roounfold_Obs(self.observable, data, response, self.config_file,
```
Comment (jdmulligan) [inline pyjetty/alice_analysis/analysis/user/substructure/run_analysis.py:245]: More funny spacings here...be sure to check the diff on github before making the PR to check for these kind of things

File: pyjetty/alice_analysis/analysis/user/substructure/run_analysis.py [hunk 4]
```diff
@@ -495,94 +579,131 @@ def plot_systematic_uncertainties(self, jetR, obs_label, obs_setting, grooming_s
...
+    # Obtain the total systematic uncertainty histograms from memory
+    min_reg_param = 3
+    reg_params = range(min_reg_param, self.max_reg_param + 1)
+    names = [ "hSystematic_Total_R{}_{}_n{}_{}-{}".format(self.utils.remove_periods(jetR), obs_label,
+                                                          reg_param, int(min_pt_truth), int(max_pt_truth))
+              for reg_param in reg_params ]
+    hSys_list = [ getattr(self, name) for name in names ]
+
+    # Obtain the statistical uncertainty histograms from disk
+    fdir = self.output_dir_main
```
Comment (jdmulligan) [inline pyjetty/alice_analysis/analysis/user/substructure/run_analysis.py:648]: A useful trick is to do `os.path.join(fdir, 'myfile.root')`

**Comment by ezradlesser** [comment]
@jdmulligan 

> * I think we should keep the ability to specify the regularization parameter manually -- e.g. take the manual value if one is specified in the config, otherwise determine it automatically (or any other implementation you prefer)

I guess this is reasonable -- though I'm not sure why someone would _want_ to do this. There isn't really a slow-down, but this can probably be changed so that a flag is passed or something in the case that someone wants to pass regularization parameters manually. 

> * There are a huge amount of whitespace changes and line breaks, etc. I'm not sure if this is intentional or not -- make sure to check the "Files changed" tab on your PR before making it and check through all of the files (the shared files at least) to make sure they include only what you need. If you in fact intentionally made all these changes, you should rather propose a formatting style you prefer and we can run a code-format commit (but it is not good to mix these with other changes).

Yes, these are intentional. In general there are many lines of code in these files which are much too long. In many code editors this means that the person reader either has to scroll horizontally to read the code, or the lines break at the end of the screen which causes somewhat random rollovers to the other edge of the screen, in either case making the code very difficult to read. I think we should follow the PEP8 guidelines here: https://www.python.org/dev/peps/pep-0008/#maximum-line-length

Another issue is that most empty lines were filled with trailing white space, which I corrected here. We should also follow the PEP8 guidelines regarding 4-space indentations instead of 2 (which are much easier to follow & read), but I have _not_ corrected this here. https://www.python.org/dev/peps/pep-0008/#indentation

There are probably some other whitespace issues which I made / didn't fix, as I was mostly just trying to make the code readable.

File: pyjetty/alice_analysis/analysis/user/substructure/run_analysis.py (hunk 4 above)
Comment (ezradlesser) [reply pyjetty/alice_analysis/analysis/user/substructure/run_analysis.py:648]: Thanks! Didn't know that took care of `'/'`

File: pyjetty/alice_analysis/analysis/user/substructure/run_analysis.py (hunk 1 above)
Comment (ezradlesser) [reply pyjetty/alice_analysis/analysis/user/substructure/run_analysis.py:12]: Okay, will try to keep these separate in the future.

**Comment by jdmulligan** [comment]
Hi @ezradlesser 

The reason to keep a manual option is twofold: (1) In general when adding new features that haven't been extensively tested it is best to keep flexibility with the old option, since there might arise cases where the automated algorithm does undesirable things (2) As you discussed the other day, it is not clear yet what will happen if e.g. there is a spike in uncertainty in a particular bin or two -- we may want to have manual control in some cases. It should take just a few lines of code to keep that option, since the functionality is already there to read the reg param from the config file (an easy way is to default to the automatic method, but check if a manual value is given in the config file and use that if so -- I think that is really <5 lines). So I would propose that you add that backwards compatibility, then I'll merge the PR. 

Regarding the code formatting, indeed it would be good to have some agreed standards (some you mention are pretty much universally agreed on, like line-length, others are more subjective, like indentation spacing). Thanks for some initiative on this -- but we should go about it differently than mixing in a large amount of format changes in other PRs. In particular: We should use a code formatting tool (e.g. [black](https://github.com/psf/black)) and do a universal commit across all of the files. And importantly: we should agree beforehand to adopt (and *maintain*) a certain convention -- code formatting is 90% communication. For this time it is ok but going forward if you want to make more than one or two formatting changes, we should do it as above.

**Comment by ezradlesser** [comment]
@jdmulligan 
One should now be able to bypass the max_reg_param implementation and use the previous implementation by just not having the max_reg_param in the config file.

File: pyjetty/alice_analysis/analysis/user/substructure/run_analysis.py (hunk 2 above)
Comment (ezradlesser) [reply pyjetty/alice_analysis/analysis/user/substructure/run_analysis.py:190]: Resolved in `c9433d0`

File: pyjetty/alice_analysis/analysis/user/substructure/run_analysis.py (hunk 4 above)
Comment (ezradlesser) [reply pyjetty/alice_analysis/analysis/user/substructure/run_analysis.py]: Resolved in `c9433d0`

**Comment by jdmulligan** [comment]
Great, thanks!!

**Pull request merged**